Сделали файл инпута

In [22]:
with open('o_henry.conllu') as f, open('input.conllu','w') as inp:
    for line in f:
        if line == '' or line == '\n':
            continue
        elif line.startswith('#'):
            inp.write(line)
        else:
            fields = [line.split('\t')[0],line.split('\t')[1]]
            for i in range(8):
                fields.append('-')
            print('\t'.join([field for field in fields]), file = inp)

In [113]:
file = open('model.tsv')
word_info = {} #{word: {tag:tag_count}}
for line in file:
    
    if line.startswith('#'):
        continue
    
    word = line.split('\t')[3].strip()
    tag = line.split('\t')[2]
    count_tag = int(line.split('\t')[1])

    #if word == '-':
        #continue

    if not word in word_info:
        word_info[word]= {}
        word_info[word][tag] = 0
    elif tag not in word_info[word]:
        word_info[word][tag] = 0
    word_info[word][tag] += count_tag

In [114]:
word_info

{'-': {'PROPN': 107,
  'NOUN': 524,
  'PUNCT': 561,
  'SCONJ': 51,
  'ADJ': 218,
  'VERB': 353,
  'ADP': 218,
  'NUM': 35,
  'PRON': 233,
  'AUX': 15,
  'CCONJ': 85,
  'DET': 67,
  'ADV': 109,
  'PART': 72,
  'INTJ': 4},
 'Русские': {'PROPN': 1, 'ADJ': 1},
 'соболя': {'NOUN': 9},
 '.': {'PUNCT': 157},
 'О.': {'PROPN': 1},
 'Генри': {'PROPN': 1},
 'Когда': {'SCONJ': 1},
 'синие': {'ADJ': 1},
 ',': {'PUNCT': 241},
 'как': {'SCONJ': 15},
 'ночь': {'NOUN': 1},
 'глаза': {'NOUN': 5},
 'Молли': {'VERB': 1, 'PROPN': 28},
 'Мак-Кивер': {'PROPN': 2},
 'положили': {'VERB': 1},
 'Малыша': {'PROPN': 4, 'NOUN': 4},
 'Брэди': {'PROPN': 7},
 'на': {'ADP': 30},
 'обе': {'NUM': 1},
 'лопатки': {'NOUN': 1},
 'он': {'PRON': 19},
 'вынужден': {'ADJ': 1},
 'был': {'AUX': 4},
 'покинуть': {'VERB': 2},
 'ряды': {'NOUN': 1},
 'банды': {'NOUN': 5},
 '«Дымовая': {'ADJ': 7},
 'труба»': {'NOUN': 7},
 'Такова': {'ADJ': 1},
 'власть': {'NOUN': 1},
 'нежных': {'ADJ': 1},
 'укоров': {'NOUN': 1},
 'подружки': {'NOUN':

In [115]:
word_tag = {}
for word, tags in word_info.items():
    word_tag[word] = max(tags, key=tags.get)

In [116]:
word_tag

{'-': 'PUNCT',
 'Русские': 'PROPN',
 'соболя': 'NOUN',
 '.': 'PUNCT',
 'О.': 'PROPN',
 'Генри': 'PROPN',
 'Когда': 'SCONJ',
 'синие': 'ADJ',
 ',': 'PUNCT',
 'как': 'SCONJ',
 'ночь': 'NOUN',
 'глаза': 'NOUN',
 'Молли': 'PROPN',
 'Мак-Кивер': 'PROPN',
 'положили': 'VERB',
 'Малыша': 'PROPN',
 'Брэди': 'PROPN',
 'на': 'ADP',
 'обе': 'NUM',
 'лопатки': 'NOUN',
 'он': 'PRON',
 'вынужден': 'ADJ',
 'был': 'AUX',
 'покинуть': 'VERB',
 'ряды': 'NOUN',
 'банды': 'NOUN',
 '«Дымовая': 'ADJ',
 'труба»': 'NOUN',
 'Такова': 'ADJ',
 'власть': 'NOUN',
 'нежных': 'ADJ',
 'укоров': 'NOUN',
 'подружки': 'NOUN',
 'и': 'CCONJ',
 'ее': 'PRON',
 'упрямого': 'ADJ',
 'пристрастия': 'NOUN',
 'к': 'ADP',
 'порядочности': 'NOUN',
 'Если': 'SCONJ',
 'эти': 'DET',
 'строки': 'NOUN',
 'прочтет': 'VERB',
 'мужчина': 'NOUN',
 'пожелаем': 'VERB',
 'ему': 'PRON',
 'испытать': 'VERB',
 'себе': 'PRON',
 'столь': 'ADV',
 'же': 'PART',
 'благотворное': 'ADJ',
 'влияние': 'NOUN',
 'завтра': 'ADV',
 'не': 'PART',
 'позднее': '

In [130]:
def tagger(line, dictionary):
    line = line.split('\t')
    word = line[1]
    tag = dictionary[word]
    line[3] = tag
    new_line = ('\t').join([field for field in line])
    return new_line

In [166]:
text = open('input.conllu')
out = open('output.conllu','w')
for line in text:
    line = line.strip()
    if line.startswith('#'):
        print(line, file = out)
    else:
        print(tagger(line, word_tag), file = out)

FileNotFoundError: [Errno 2] No such file or directory: 'input.conllu'

In [165]:
with open('o_henry.conllu') as inp, open('output.conllu') as out:
    mistakes = 0
    word_tag_inp = [(line.split('\t')[1], line.split('\t')[3]) for line in inp if not line.startswith('#') and line != '\n']
    tags_out = [line.split('\t')[3] for line in out if not line.startswith('#') and line != '\n']
    total = len(tags_out)
    for i in range(total):
        if word_tag_inp[i][1] != tags_out[i]:
            mistakes +=1
            print(word_tag_inp[i][0], tags_inp[i], tags_out[i], sep = '\t')
    print((total-mistakes)*100/total)

Молли	VERB	PROPN
и	PART	CCONJ
что	PRON	SCONJ
Да	CCONJ	PART
все	PRON	DET
и	PART	CCONJ
ни	CCONJ	PART
Русские	ADJ	PROPN
что	PRON	SCONJ
щеки	ADV	NOUN
да	CCONJ	PART
хорошо	ADV	ADJ
русских	NOUN	ADJ
все	PART	DET
Рэнсом	NOUN	PROPN
и	PART	CCONJ
Рэнсом	NOUN	PROPN
что	PRON	SCONJ
все	PRON	DET
что	PRON	SCONJ
все	PRON	DET
все	PRON	DET
Малыша	NOUN	PROPN
пусть	PART	SCONJ
Малыша	NOUN	PROPN
полицейский	NOUN	ADJ
полицейского	NOUN	ADJ
чем	SCONJ	PRON
Малыша	NOUN	PROPN
Малыша	NOUN	PROPN
98.86749716874291


1. Accuracy = 98.9 %
2. We can see some mistakes in original tagging) for example, 'Молли' should be PROPN
3. In the unigram model we can store such single-word features in the unigram model as Zaliznjak code, semantic class, etc